aspnetcore/blazor/webassembly-lazy-load-assemblies.md

```diff
@@ -0,0 +1,213 @@
+
+Inside `OnNavigateAsync`, implement logic to determine the assemblies to load. Options include:
+
+* Conditional checks inside the `OnNavigateAsync` method.
+* A lookup table that maps routes to assembly names, either injected into the component or implemented within the [`@code`](xref:mvc/views/razor#code) block.
+
+`LazyAssemblyLoader` is a framework-provided singleton service for loading assemblies. Inject `LazyAssemblyLoader` into the `Router` component:
+
+```razor
+...
+@using Microsoft.AspNetCore.Components.WebAssembly.Services
+@inject LazyAssemblyLoader assemblyLoader
+
+...
+```
+
+The `LazyAssemblyLoader` provides the `LoadAssembliesAsync` method that:
+
+* Uses JS interop to fetch assemblies via a network call.
+* Loads assemblies into the runtime executing on WebAssembly in the browser.
+
+> [!NOTE]
+> The framework's lazy loading implementation supports prerendering on the server. During prerendering, all assemblies, including those marked for lazy loading, are assumed to be loaded.
+
+### User interaction with `<Navigating>` content
+
+While loading assemblies, which can take several seconds, the `Router` component can indicate to the user that a page transition is occurring:
+
+* Add an [`@using`](xref:mvc/views/razor#using) directive for the <xref:Microsoft.AspNetCore.Components.Routing?displayProperty=fullName> namespace.
+* Add a `<Navigating>` tag to the component with markup to display during page transition events.
+
+```razor
+...
+@using Microsoft.AspNetCore.Components.Routing
+...
+
+<Router ...>
+    <Navigating>
+        <div style="...">
+            <p>Loading the requested page&hellip;</p>
+        </div>
+    </Navigating>
+</Router>
+
+...
+```
+
+### Handle cancellations in `OnNavigateAsync`
+
+The `NavigationContext` object passed to the `OnNavigateAsync` callback contains a `CancellationToken` that's set when a new navigation event occurs. The `OnNavigateAsync` callback must throw when this cancellation token is set to avoid continuing to run the `OnNavigateAsync` callback on a outdated navigation.
+
+If a user navigates to Route A and then immediately to Route B, the app shouldn't continue running the `OnNavigateAsync` callback for Route A:
+
+```razor
+@inject HttpClient Http
+@inject ProductCatalog Products
+
+<Router AppAssembly="@typeof(Program).Assembly" 
+    OnNavigateAsync="@OnNavigateAsync">
+    ...
+</Router>
+
+@code {
+    private async Task OnNavigateAsync(NavigationContext args)
+    {
+        if (args.Path == "/about") 
+        {
+            var stats = new Stats = { Page = "/about" };
+            await Http.PostAsJsonAsync("api/visited", stats, args.CancellationToken);
+        }
+        else if (args.Path == "/store")
+        {
+            var productIds = [345, 789, 135, 689];
+            foreach (var productId in productIds) 
+            {
+                args.CancellationToken.ThrowIfCancellationRequested();
```

Comment:
This is like a API thing. Usually the cancelation token ends up being part of the async method signature i.e. it's more common to see: `OnNavigateAsync(NavigationContext context, CancelationToken cts)`. Nothing to do here, just something that came to me when I saw this.